rewrite-test/src/main/kotlin/org/openrewrite/java/cleanup/CombineSemanticallyEqualCatchBlocksTest.kt

```diff
@@ -0,0 +1,264 @@
+                void method() {
+                    try {
+                    } catch (IllegalArgumentException ex) {
+                        String msg = "message";
+                        throw new RuntimeException(msg, ex);
+                    } catch (IllegalStateException ex) {
+                        String msg = "different message";
+                        throw new RuntimeException(msg, ex);
+                    }
+                }
+            }
+        """)
+    )
+
+    @Test
+    fun createNewMultiCatch(jp: JavaParser) = rewriteRun(
+        {spec -> spec.parser(jp)},
+        java("""
+            class Test {
+                void method0() {
+                    try {
+                    } catch (IllegalArgumentException ex) {
+                        String msg = "combine";
+                        throw new RuntimeException(msg, ex);
+                    } catch (IllegalStateException ex) {
+                        String msg = "combine";
+                        throw new RuntimeException(msg, ex);
+                    }
+                }
+            }
+        """,
+        """
+            class Test {
+                void method0() {
+                    try {
+                    } catch (IllegalArgumentException | IllegalStateException ex) {
+                        String msg = "combine";
+                        throw new RuntimeException(msg, ex);
+                    }
+                }
+            }
+        """)
+    )
+
+    @Test
+    fun multiCatchExists(jp: JavaParser) = rewriteRun(
+        {spec -> spec.parser(jp)},
+        java("""
+            class Test {
+                void method() {
+                    try {
+                    } catch (IllegalArgumentException | IllegalStateException ex) {
+                        checkEquivalence("same");
+                        throw new RuntimeException(ex);
+                    } catch (RuntimeException ex) {
+                        checkEquivalence("diff");
+                        throw new RuntimeException(ex);
+                    } catch (IllegalAccessException ex) {
+                        checkEquivalence("same");
+                        throw new RuntimeException(ex);
+                    }
+                }
+                void checkEquivalence(String s) {
+                }
+            }
+        """,
+        """
+            class Test {
+                void method() {
+                    try {
+                    } catch (IllegalArgumentException | IllegalStateException | IllegalAccessException ex) {
+                        checkEquivalence("same");
+                        throw new RuntimeException(ex);
+                    } catch (RuntimeException ex) {
+                        checkEquivalence("diff");
+                        throw new RuntimeException(ex);
+                    }
+                }
+                void checkEquivalence(String s) {
+                }
+            }
+        """)
```

Comment:
Changing the order the catch blocks are evaluated in isn't always safe and can represent a change in program semantics. Catch blocks are evaluated in order and the first matching block is taken. 

Given exception types `A`, `B`, `C` caught in that order, it is safe to transform into `A | C`, `B` _only_ if `C` is not a subtype/implementation of `B` and therefore could never be handled by that catch.

Reply:
That's fair, I'll update the qualification combine catch blocks to consider the previous caught types.